### ISSUE TITLE
[META] Questions about .NET Core Support

### ISSUE THREAD

**Issue by shiftkey** [octokit/octokit.net#1115]
This is a tracking issue for questions related to https://github.com/octokit/octokit.net/issues/832.

.NET Core is currently under development (and the next milestone RC2 has a bunch of changes which will likely impact the tooling side of things) so there might be some extra things to consider here.

**Comment by M-Zuber** [comment]
Do we really want to go through the port now? 
Wait a week or two and see if they come up with a a date and based on that decide if we will wait for RC2 to start porting? 
If we are starting now I would say the first thing to figure out is nuget package creation. 
Say pick two random profiles from the list we want to support and see what it takes to get the build to produce the packages correctly

**Comment by shana** [comment]
> If we are starting now I would say the first thing to figure out is nuget package creation. 

@M-Zuber Just fyi, the GSoC student proposal application period starts March 14 and ends March 29, and the official coding period for the program doesn't start before May 29. "Now" is quite a ways away for GSoC :smile:

**Comment by shana** [comment]
#### GSoC Students

Make sure you read https://github.com/github/mentorships/issues/100 and http://write.flossmanuals.net/gsocstudentguide/what-is-google-summer-of-code/ and understand how the program works, how proposals are submitted, and all the milestones involved.

**Comment by M-Zuber** [comment]
>Just fyi, the GSoC student proposal application period starts March 14 and ends March 29, and the official coding period for the program doesn't start before May 29. "Now" is quite a ways away for GSoC :smile:

I missed that connection :grinning:

**Comment by prayankmathur** [comment]
Hello developers

I am Prayank Mathur, currently pursuing my final yr in Computer Science & Engineering from Indian School of Mines, Dhanbad. 

I have had previous experience working in the C#  .NET framework where i developed an automation tool for Microsoft IDC during my summer internship last year. The main aim was to create an automation tool that can automate the testing of the daily builds of Word Android app on various android devices.

I also developed an Employee Database Management System for Tata Steel in vb using the .NET framework during  summer'14.

I have done 2 mini projects on C# .NET framework too.

https://github.com/prayankmathur/

I am really interested in the octokit project and would like to get started on the same. Please suggest me with some warm up tasks. I have read the above mentioned document.

Thanks
Regards
Prayank Mathur

**Comment by shiftkey** [comment]
@prayankmathur we have a few outstanding tasks marked [easy-fix](https://github.com/octokit/octokit.net/issues?q=is%3Aissue+is%3Aopen+label%3Aeasy-fix) - please check them out!

**Comment by shiftkey** [comment]
Someone asked me in another channel to summarize .NET Core in relation to the .NET Framework. Here's a quick brain dump:

 - .NET Core is a subset of the .NET Framework which runs on Windows, Linux,
   OSX and other operating systems
 - The [CoreCLR](https://github.com/dotnet/coreclr/) (the runtime for .NET Core)
   is an open source repository
 - Core libraries for .NET Core is distributed via NuGet packages - the 
   [CoreFX](https://github.com/dotnet/corefx/) repository has all the source for
   the .NET Core libraries
 - Various APIs have not been ported over to .NET Core
   - features which are Windows-specific - e.g. WPF, WinForms
   - features which are not suitable for their v1 target of running webapps
   - APIs which are not intended to be supported long-term
 - The other part of this project the [dotnet CLI](https://github.com/dotnet/cli/),
   which allows you to build and run programs which target .NET Core from the
   command line - this is just now available to test, but not quite v1

This is different to the Portable Class Library targets, which were designed to
simplify the platforms you targeted, but were limited to where .NET was already
running - e.g. Mono/Xamarin/.NET Framework

Octokit.net already has a PCL profile that's suitable to migrate to .NET Core,
so most of the work around this will likely involve migrating to the
[project.json](https://github.com/aspnet/Home/wiki/Project.json-file) format to
define the dependencies and platforms that Octokit needs to support, and also
testing to verify that there are no unexpected issues with running on .NET Core
so that we're ready to be used when .NET Core 1.0 RTMs

**Comment by devkhan** [comment]
@shiftkey What parts of the project we will be needing to re-write to make this .NET Core compatible? I mean, how much code is there which is .NET Framework specific and not available in .NET Core or is never going to be?

**Comment by shiftkey** [comment]
@devkhan currently we juggle different csproj files for each target. I think a bulk of the work will be in porting those over to a unified project.json file - this will then allow us to add the new targets for the various [.NET Platform Standards](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md) available around .NET Core and verify that these can be built (we might need to introduce some `#ifdef` but I think that's minor).

After that, we'd need to look at packaging as `project.json` can also be used to create packages "for free" - but we do some customization like incorporate [Linqpad](https://www.linqpad.net/) samples and [SourceLink](https://github.com/ctaggart/sourcelink) the PDBs so the source can be stepped through easily.

**Comment by shiftkey** [comment]
> And it will generate different NuGet packages for different targets?

One package (with multiple targets inside) is generated from a project.json file - hence the unification stuff. But this also raises headaches as we have been building the package from hand before.

> Also, the Mono and Xamarin ones will still be kept as different projects or they are going to be dissolved in favour of the X-platform .NET Core? 

They should stay around - the platforms are mentioned in the .NET Platform Standards document linked to above, but I've not explored this in practice.

> I read a [blog](https://oren.codes/2015/06/16/demystifying-pcls-net-core-dnx-and-uwp-redux/), which said that setting dotnet as target in project.json means we are specifying that our project is compatible with any platform our dependencies are compatible. So, does this mean we will have to make sure we have no dependencies which are not available on .NET Core?

Kind of. `dotnet` was the name intended for to represent various API sets - to simplify targetting various flavours of .NET. That article is now not accurate - these are now named `netstandard`, but the details are largely the same.

How this looks in practice is that you specify the _version_ of netstandard inside a project.json file, like this:

```
{
    // other stuff here
    "frameworks": {
        "net40": {
           // other stuff here 
        },
        "net45": {
           // other stuff here 
        },
        "net46": {
           // other stuff here 
        },
        "netstandard1.0": {
           // other stuff here 
        },
        "netstandard1.3": {
           // other stuff here 
        },
        "sl5": {
           // other stuff here 
        }
    }
}
```

The new `dotnet` CLI tools will then try and build your code for each of the `framework` entries defined, and the build outputs become the contents of the NuGet package...

**Comment by devkhan** [comment]
@shiftkey Thanks for clearing that out. :+1: One question though, from [here](http://dotnet.readthedocs.org/en/latest/concepts/class-libraries.html#net-core-class-libraries), .NET Core (netstandard) class libs are already supported on some of these platforms, e.g. net45, net46, win8, etc. So, will they also be continued or merged into one?

**Comment by shiftkey** [comment]
@devkhan I'd like to keep the existing platforms around for now - there may be some platform-specific features that aren't present in the `netstandard` equivalent. I'm involved with another project which is undergoing that process, so I definitely believe it's acheivable.

**Comment by prayankmathur** [comment]
> currently we juggle different csproj files for each target. I think a bulk of the work will be in porting those over to a unified project.json file - this will then allow us to add the new targets for the various .NET Platform Standards available around .NET Core

@shiftkey 

How exactly do we port a csproj file to project.json
Do we have to do it manually defining each and every dependency as mentioned [here](https://github.com/aspnet/Home/wiki/Project.json-file) or is there a tool available ?

And once that is done, we then need to modify the build.fsx file to incorporate the changes ?
And after that see what all errors creep in and try to rectify it using #if 

>After that, we'd need to look at packaging as project.json can also be used to create packages "for free"
You mean Nuget packages right ?

**Comment by devkhan** [comment]
@prayankmathur I guess it is done by hand right now. But there must be tools available.

Maybe we will have a unified project.josn which will target the current .NET Framework, .Net45 and PCL, which all have different projects right now. And the project.json will just contain definitions for different targets (their dependencies). You can take a look at the [`target-the-coreclr`](https://github.com/octokit/octokit.net/tree/target-the-coreclr) branch for reference.

@shiftkey can confirm.

**Comment by shiftkey** [comment]
@prayankmathur

> How exactly do we port a csproj file to project.json

There are no tools to convert a `csproj` over to `project.json`, but it's easy to add a project.json (and the `.xproj`) alongside the existing `.csproj` to help with the transition. An earlier example of this can be seen [here](https://github.com/octokit/octokit.net/tree/target-the-coreclr/Octokit) although that's rather out of date.

> Do we have to do it manually defining each and every dependency as mentioned here or is there a tool available ?

One of the benefits of this new approach is that you can define a `dependencies` array that will use packages (or local projects) which are applied to all `frameworks` - instead of duplicating entries.

> And once that is done, we then need to modify the build.fsx file to incorporate the changes ?

We may not need to change much here. If not, we'll need to look at using the [`dotnet` CLI](https://github.com/dotnet/cli) tools here... 

> You mean Nuget packages right ?

Yes

**Comment by shiftkey** [comment]
@devkhan exactly - one `project.json` file can target many frameworks, and can replace the multiple `csproj` files we currently maintain. Because the project.json file supports wildcard expressions to include code, like this:

```
{
    "compile": "*.cs",
    "exclude": "buggy/**/*.cs",
    "publishExclude": "something/**/*.*",
    "resource": "embed/**/*.*"
}
```

We can drop the `.\build FixProjects` script we have to ensure all the `csproj` files are in sync - and no more merge conflicts from `csproj` files!!!

**Comment by prayankmathur** [comment]
@shiftkey

Do you believe this task is gonna take 2 months if we only have to do things that is discussed in this meta issue.

I think we can finish it fast once everything is figured out step by step on how to do it .

Am i right or this is gonna take time because i haven't have any experience in writing a project. json from .csproj ?

**Comment by shiftkey** [comment]
@prayankmathur 2 months of part-time work (this could just be a few hours each week), including integrating this into the process, shipping preview packages and testing, plus allowing for unexpected upstream issues that might occur with pre-release packages feel about right...

**Comment by ikourfaln** [comment]
Hi,
OK, as I said before (#1257), CoreCLR has a functioning HttpClient (System.Net.Http), it is included in .NET 4.5 and .NET Core, so it will work on ASP.NET Core.
Use System.Net.Http.HttpClient to communicate over HTTP with GitHub web services. And you need to remove Microsoft.Net.Http (whitch it depend to two other packages :wink: ).

**Comment by ikourfaln** [comment]
So @shiftkey. whats the status of this issue ?!

**Comment by shiftkey** [comment]
@ikourfaln I've been experimenting with .NET Core for the upcoming RC2 bits (not on this project, on something more complex). My current focus is wrapping up the `ApiOptions` release, and by the time that's done I think the RC2 tooling bits like `dotnet cli` should be ready for mainstream.

However I've got way too much on my plate these days (and over the coming months) and none of it needs .NET Core support, so I'd love it if someone stepped up for this based on needing it for their usage...

**Comment by alfhenrik** [comment]
As I'm about to embark on a new .NET Core side project in which I'd (most likely) like to use Octokit I can have a look into this one

**Comment by shiftkey** [comment]
@alfhenrik now that Rx.NET preview packages supporting .NET Core are available - [details here](https://github.com/Reactive-Extensions/Rx.NET/issues/148#issuecomment-219506305) - this is now something we can start exploring. You're more than welcome to dive in!

**Comment by radu-matei** [comment]
I don't think that converting to `project.json` is going to be required, since all .NET projects will use the `.csproj` file.

> After looking at our choices, it was apparent that it would be easier to move .NET Core projects to .csproj/MSBuild so all .NET projects use the same tooling and build system. [(More on this here.)](https://blogs.msdn.microsoft.com/dotnet/2016/05/23/changes-to-project-json/)

In a couple of weeks the RTM will be released, so I think it would be a great moment to start working on porting to .NET Core.

What do you people think?

**Comment by shiftkey** [comment]
Okay, so now I can finally talk about porting Rx.NET to .NET Core, have a look at our setup for one of the projects:

https://github.com/Reactive-Extensions/Rx.NET/blob/master/Rx.NET/Source/System.Reactive.Core/project.json

Feel free to repurpose this to suit - these bits are ready for GA in a couple of weeks.

**Comment by alfhenrik** [comment]
I'm sorry to say it but, I thought I was in a position to have a crack at this, but it turns out I'm not...so I'm walking away from it. Thought I'd just mention it here so if someone else wanted to pick it up they don't need to confirm the status from my side.

**Comment by shiftkey** [comment]
Closing this out in favour of #1419 which has more information and is more recent.
